src/LmdbCache.ts

```diff
@@ -95,43 +102,56 @@ export class LmdbCache<V = any> implements SortKeyCache<V> {
 
   async put(cacheKey: CacheKey, value: V): Promise<void> {
     return this.db.childTransaction(() => {
-      this.db.put(`${cacheKey.key}|${cacheKey.sortKey}`, value);
-
-      // Get number of elements that is already in cache.
-      // +1 to account for the element we just put and will be inserted with this transaction
-      const numInCache =
-        1 +
-        this.db.getKeysCount({
-          start: `${cacheKey.key}|${genesisSortKey}`,
-          end: `${cacheKey.key}|${cacheKey.sortKey}`
-        });
+      this.barePut(cacheKey, value);
+      this.removeOldestEntries(cacheKey);
+    });
+  }
 
-      // Make sure there isn't too many entries for one contract
-      if (numInCache <= this.lmdbOptions.maxEntriesPerContract) {
-        // We're below the limit, finish
-        return;
-      }
+  async barePut(cacheKey: CacheKey, value: V): Promise<boolean> {
+    return this.db.put(`${cacheKey.key}|${cacheKey.sortKey}`, value);
+  }
 
-      // Remove the oldest entries, so after the final put there's minEntriesPerContract present
-      const numToRemove = numInCache - this.lmdbOptions.minEntriesPerContract;
-      // Remove entries one by one, it's in a transaction so changes will be applied all at once
-      this.db
-        .getKeys({
-          start: `${cacheKey.key}|${genesisSortKey}`,
-          limit: numToRemove
-        })
-        .forEach((key) => {
-          this.db.remove(key);
-        });
-    });
+  async removeOldestEntries(cacheKey: CacheKey) {
+    // Get number of elements that is already in cache.
+    // +1 to account for the element we just put and will be inserted with this transaction
+    const numInCache =
+      1 +
+      this.db.getKeysCount({
+        start: `${cacheKey.key}|${genesisSortKey}`,
+        end: `${cacheKey.key}|${cacheKey.sortKey}`
+      });
+
+    // Make sure there isn't too many entries for one contract
+    if (numInCache <= this.lmdbOptions.maxEntriesPerContract) {
+      // We're below the limit, finish
+      return;
+    }
+
+    // Remove the oldest entries, so after the final put there's minEntriesPerContract present
+    const numToRemove = numInCache - this.lmdbOptions.minEntriesPerContract;
+    // Remove entries one by one, it's in a transaction so changes will be applied all at once
```

Comment:
well, now there's no guarantee the it will be in a transaction, as it depends on calling code?

Reply:
yes, that's correct, let's make it private